src/mds/MDLog.cc

```diff
@@ -416,65 +416,72 @@ void MDLog::_submit_thread()
     locker.unlock();
 
     if (data.le) {
-      LogEvent *le = data.le;
+      // will be destroyed when this scope ends.
+      std::unique_ptr<LogEvent> le{data.le};
       LogSegment *ls = le->_segment;
+
+      const uint64_t start_offset = journaler->get_write_pos();
+      le->set_start_off(start_offset);
+
       // encode it, with event type
       bufferlist bl;
       le->encode_with_header(bl, features);
 
-      uint64_t write_pos = journaler->get_write_pos();
-
-      le->set_start_off(write_pos);
-      if (dynamic_cast<SegmentBoundary*>(le)) {
-	ls->offset = write_pos;
-      }
-
       if (bl.length() >= event_large_threshold.load()) {
         dout(5) << "large event detected!" << dendl;
         logger->inc(l_mdl_evlrg);
       }
 
-      dout(5) << "_submit_thread " << write_pos << "~" << bl.length()
+      dout(5) << "_submit_thread " << start_offset << "~" << bl.length()
 	      << " : " << *le << dendl;
 
       // journal it.
-      const uint64_t new_write_pos = journaler->append_entry(bl);  // bl is destroyed.
-      ls->end = new_write_pos;
+      const uint64_t end_offset = journaler->append_entry(bl); // bl is destroyed.
 
-      MDSLogContextBase *fin;
       if (data.fin) {
-	fin = dynamic_cast<MDSLogContextBase*>(data.fin);
-	ceph_assert(fin);
-	fin->set_write_pos(new_write_pos);
+	MDSLogContextBase* mds_fin = dynamic_cast<MDSLogContextBase*>(data.fin);
+        ceph_assert(mds_fin);
+        mds_fin->set_write_pos(end_offset);
       } else {
-	fin = new C_MDL_Flushed(this, new_write_pos);
+        // we should always have an MDSLogContextBase
+        // completion handler for serialized events
+        // because we update the safe_pos when it's finished
+        data.fin = new C_MDL_Flushed(this, end_offset);
       }
 
-      journaler->wait_for_flush(fin);
-
-      if (data.flush)
-	journaler->flush();
+      if (logger) {
+	logger->set(l_mdl_wrpos, end_offset);
+      }
 
-      if (logger)
-	logger->set(l_mdl_wrpos, ls->end);
+      // acquire the mds lock until the end of the scope
+      // to update the LS offsets and major offsets
+      std::lock_guard l(mds->mds_lock);
```

Comment:
I still need to go through these changes in more detail but this will not fly. The `mds_lock` has too much thread contention already. Also, the submit thread is intended to operate completely outside the `mds_lock`. I do not think we should ever change that.

I believe there may be another way to protect the LogSegment members behind the `mds_lock`: add the `LogSegment` to `C_MDL_Flushed` and update the segment's end/offset in the completion (`::finish`).

Reply:
C_MDL_Flushed will take mds_lock, so not much difference in terms of contention. But I agree that we could benefit from performing the update on the finisher thread instead of the submit thread.

Reply:
As you said, the journaler thread already regularly acquires the `mds_lock` so it's advantageous to use it for this purpose instead of adding a new thread to the contention.

Reply:
Please see a1c9d1a8be08807d83eac0c7bf75cfe04101768c, I removed the need for locking.